### PR TITLE
refactor code generator format

### PR DESCRIPTION
**Pull request by qinzuoyan**
- refactor code generator format
- fix failure_detector bug: _is_master is not checked
- fix network bug: _matcher should not be null in both server and client session

### CODE REVIEW

**Comment by qinzuoyan** [comment]
@imzhenyu , please review it

**Comment by imzhenyu** [comment]
@qinzuoyan why matcher is needed on server session?

File: src/dist/failure_detector/failure_detector.cpp [hunk 1]
```diff
@@ -417,6 +417,13 @@ bool failure_detector::end_ping_internal(::dsn::error_code err, const beacon_ack
         return true;
     }
 
+    if (!ack.is_master)
+    {
+        dwarn("remote node[%s] is not master, ack.primary_node[%s]",
+              node.to_string(), ack.primary_node.to_string());
+        return true;
+    }
+
```
Comment (shengofsun) [inline src/dist/failure_detector/failure_detector.cpp:426]: "is_master" shouldn't be handled like this for better performance when switching master.

**Comment by qinzuoyan** [comment]
@imzhenyu , because rpc_session::on_recv_reply() can be called in server session, when this session received a forwarded reply.

**Comment by imzhenyu** [comment]
Forwarding a request does not expect a reply to this machine but the original machine who initiates the rpc. Do you encounter this scenario during test?

**Comment by imzhenyu** [comment]
Just realize you are talking about forwarding a reply, which should not happen in our system as replies can always send to the initiation side directly if a forwarding is possible. We may add an assertion there to avoid misuse though.

**Comment by qinzuoyan** [comment]
A request to server B, but B forward the request to server C, C then reply to server A, as in rpc_engine::reply(), C send response message to A (where C is client session and A is server session), in this case, A's server session will receive reply message from C, and need matcher.

**Comment by imzhenyu** [comment]
I see. I did not test this setting before though the code logic is implemented there ... My bad.

**Comment by qinzuoyan** [comment]
So finally, do we need matcher in server session?

**Comment by imzhenyu** [comment]
Sure, let's having it in server session as well.

**Comment by qinzuoyan** [comment]
there are some other problems, I am looking into the code.

**Comment by qinzuoyan** [comment]
@imzhenyu , as fixed above, there is a group address leader change bug, and you'd better pay attention to new TODOs.

**Comment by qinzuoyan** [comment]
@imzhenyu , I think the rpc forwarding really make things complicated, and is it really worthy?

**Comment by qinzuoyan** [comment]
and, another problem is: if the forwarding target (node C) is dead, then the forwarding request is discarded, and the original client will wait util timeout.

now general call to meta servers use rpc forwarding (refer to meta_service::check_primary()), but failure detector heartbeat do not use rpc forwarding (refer to meta_server_failure_detector::on_ping()), which will cause inconsistency.

**Comment by imzhenyu** [comment]
@qinzuoyan I don't get the necessity of introducing "from_primary_address". It seems I do make a mistake where I thought from_address is sent through network but in fact always assigned locally. We can put from_address into the header and then the forwarded peers (no matter how many hops) will remember the original site. In that case, I think "from_primary_address" is unnecessary.

**Comment by imzhenyu** [comment]
As for FD, it is kinds of tricky and unless we can reason the effect crystal clear, I think we should remain the current practice, but not use group address as target address. I am not sure how such inconsistency will cause problem.

**Comment by imzhenyu** [comment]
I got how the auto-changed group address may break the assertion. As long as we use `_meta_servers.leader()` instead of `_meta_servers` itself as the ping target, there should not be an issue.

**Comment by imzhenyu** [comment]
@qinzuoyan You may rollback your last commit https://github.com/qinzuoyan/rDSN/commit/f88e2043b95aebb8a0f7706fdfc05cd203114c90 and merge the changes from #267 , then the issue should be fine.

**Comment by imzhenyu** [comment]
@qinzuoyan  replica1.THREAD_POOL_REPLICATION1.050002ac00010004: replica.stub:429:dsn::replication::replica_stub::on_config_proposal(): on_config_proposal: not connected, ignore

Just find lots of such errors this morning - it may be because of the FD change made earlier. We need to merge your quick fix asap.

Thx,
Zhenyu

**Comment by imzhenyu** [comment]
This seems unrelated - it is due to my previous rpc forwarding change, as now we cannot use _is_client to check whether an incoming msg is a request or not, but use header.context.u.is_request. I'll come up a fix pr quickly for this.

**Comment by qinzuoyan** [comment]
there is a better fix, I am working on it, wait later

**Comment by qinzuoyan** [comment]
@imzhenyu, actually, there are still some bugsi related to rpc forwarding, and I am wording to fix it.

**Comment by imzhenyu** [comment]
@qinzuoyan Great. I've just come up a fix at https://github.com/imzhenyu/rDSN/commit/e9803b0dc6ec1871825b19cb1430cf1109a81563 
It fixes the above problem. But no worry, I'll wait your pr first in that case.

**Comment by qinzuoyan** [comment]
@imzhenyu , fixed as above, please review it. Main points:
- add comments/TODOs for message_header::from_address and message_ex::to_address
- add TODOs in rpc_client_matcher::on_recv_reply()
- add rpc_group_address::_update_leader_on_rpc_forward
- fix inject_drop_message()
- fix bug of re-sealing the message when header.is_forwarded is changed
- improve on_rpc_request_enqueue() to return bool, and add injection in on_recv_request()
- add unit-test of group_address_change_leader in dsn.core.tests

**Comment by imzhenyu** [comment]
In general, the change is great! Regarding to the new Boolean joinpoint, I think it is fine to add it, though the cost of adding io-session to each reply msg is worth considering in the future.

**Comment by qinzuoyan** [comment]
Yes, I think a ptr assignment is acceptable, considering it makes world clean (not leaving a special case).

**Comment by imzhenyu** [comment]
Agree, let's do the opt only when it surfaces.

**Pull request merged**